### PR TITLE
[SPARK-11740][Streaming]Fix the race condition of two checkpoints in a batch

### PR DESCRIPTION
**Pull request by zsxwing**
We will do checkpoint when generating a batch and completing a batch. When the processing time of a batch is greater than the batch interval, checkpointing for completing an old batch may run after checkpointing for generating a new batch. If this happens, checkpoint of an old batch actually has the latest information, so we want to recovery from it. This PR will use the latest checkpoint time as the file name, so that we can always recovery from the latest checkpoint file.

### CODE REVIEW

**Comment by zsxwing** [comment]
retest this please

File: python/pyspark/streaming/tests.py [hunk 1]
```diff
@@ -699,11 +698,11 @@ def check_output(n):
         # Verify that getOrCreate() uses existing SparkContext
         self.ssc.stop(True, True)
         time.sleep(1)
-        sc = SparkContext(SparkConf())
```
Comment (tdas) [inline python/pyspark/streaming/tests.py]: is this change needed for this PR?

**Comment by tdas** [comment]
LGTM.
I wonder whether we can add a unit test to test this behavior. We could create a CheckpointWriter with a mock JobGenerator, and then test whether the latest checkpoint time is respected. 

BTW, I would like to merge this to older branches like 1.5 and 1.4 as well. 

@harishreedharan Can you also take a look at this?

File: python/pyspark/streaming/tests.py (hunk 1 above)
Comment (zsxwing) [inline python/pyspark/streaming/tests.py]: Just to make sure `tearDown` can shutdown it if there is any exception.

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala [hunk 2]
```diff
@@ -187,16 +187,27 @@ class CheckpointWriter(
...
       val tempFile = new Path(checkpointDir, "temp")
-      val checkpointFile = Checkpoint.checkpointFile(checkpointDir, checkpointTime)
-      val backupFile = Checkpoint.checkpointBackupFile(checkpointDir, checkpointTime)
+      // We will do checkpoint when generating a batch and completing a batch. When the processing
+      // time of a batch is greater than the batch interval, checkpointing for completing an old
+      // batch may run after checkpointing of a new batch. If this happens, checkpoint of an old
+      // batch actually has the latest information, so we want to recovery from it. Therefore, we
+      // also use the latest checkpoint time as the file name, so that we can recovery from the
+      // latest checkpoint file.
+      val checkpointFile = Checkpoint.checkpointFile(checkpointDir, latestCheckpointTime)
```
Comment (harishreedharan) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: So the idea here is that if an older batch's completion-checkpoint comes in after a new batch's initial-checkpoint, we overwrite the initial checkpoint (since we would not reset the `latestCheckpointTime`)? 

This actually could essentially mean two checkpoints being written to the same files.

Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: > This actually could essentially mean two checkpoints being written to the same files.

The previous file will be renamed to `backupFile`.

Comment (harishreedharan) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: I don't think you get what I am saying. I am saying that two threads could run at the same time writing out data to the exact same files. 

If I am not mistaken, there is a bug here that could lead to 2 checkpoints running at the same time, writing to the same files.
-- Checkpoint 1: Completion of Batch Time t
-- Checkpoint 2: Start of Batch Time t+1

Checkpoint 2 starts -> `latestCheckpoint = t + 1`
Checkpoint 1 starts -> since `latestCheckpoint != null` and `latestCheckpoint > checkpointTime`, we would not reset `latestCheckpoint`, so both checkpoints would use the same file name to write their checkpoints out.

Because of this, depending on which thread reaches the tempFile creation first, that would win - which is non-deterministic. The other thread would end up hitting an exception.

Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: There is only one thread writing the checkpoint file. See `val executor = Executors.newFixedThreadPool(1)` in `CheckpointWriter`.

Comment (harishreedharan) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: Ok, then we are fine. Can you put in a comment where the executor is being created, so we don't end up causing a bug due to this class not being thread-safe.

**Comment by harishreedharan** [comment]
LGTM. Thanks @zsxwing !

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala (hunk 2 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: Updated

**Comment by zsxwing** [comment]
Added a unit test using mock

**Comment by tdas** [comment]
Thanks @zsxwing and @harishreedharan. Merging this to master, 1.6 and 1.5
